source/OptimaJet.DWKit.StarterApplication/Forms/Projects/CreateForm.cs

```diff
@@ -0,0 +1,94 @@
+using JsonApiDotNetCore.Data;
+using JsonApiDotNetCore.Internal;
+using Microsoft.EntityFrameworkCore;
+using Optimajet.DWKit.StarterApplication.Models;
+using OptimaJet.DWKit.StarterApplication.Repositories;
+using OptimaJet.DWKit.StarterApplication.Services;
+using Serilog;
+
+namespace OptimaJet.DWKit.StarterApplication.Forms.Projects
+{
+    public class CreateForm : BaseForm
+    {
+        public UserRepository UserRepository { get; set; }
+        public GroupRepository GroupRepository { get; set; }
+        public IEntityRepository<Organization> OrganizationRepository { get; set; }
+        public ICurrentUserContext CurrentUserContext { get; }
+
+        public CreateForm(
+            UserRepository userRepository,
+            GroupRepository groupRepository,
+            ICurrentUserContext currentUserContext,
+            IEntityRepository<Organization> organizationRepository)
+        {
+            UserRepository = userRepository;
+            GroupRepository = groupRepository;
+            OrganizationRepository = organizationRepository;
+            CurrentUserContext = currentUserContext;
+        }
+
+        public bool IsValid(Project project)
+        {
+            // If these fields aren't filled in, then let the foreign key failure 
+            // be reported
+            var retVal = true;
+            if ((project.OrganizationId != 0)
+                && (project.OwnerId != 0)
+                && (project.GroupId != 0))
+            {
+                var organization = OrganizationRepository.Get()
+                        .Where(o => o.Id == project.OrganizationId)
+                        .FirstOrDefaultAsync().Result;
+                var group = GroupRepository.Get()
+                       .Where(g => g.Id == project.GroupId)
+                       .Include(g => g.Owner).FirstOrDefaultAsync().Result;
+                var owner = UserRepository.Get()
+                        .Where(u => u.Id == project.OwnerId)
+                        .Include(u => u.GroupMemberships)
+                            .ThenInclude(gm => gm.Group)
+                        .Include(u => u.OrganizationMemberships)
+                            .ThenInclude(om => om.Organization)
+                        .FirstOrDefaultAsync().Result;
+
+                if ((organization == null) || (group == null) || (owner == null))
+                {
+                    // Allowing it to return in these cases should cause the base to hit normal
+                    // foreign key failures
+                    var message = $"Project '{project.Name}': Organization Group or Owner not present or invalid";
+                    Log.Error(message);
+                    return true;
+                }
+                if (organization != group.Owner)
+                {
+                    retVal = false;
+                    var message = $"Project '{project.Name}': Group '{group.Name}' not owned by project organization '{organization.Name}'";
```

Comment:
all this error construction logic should be extracted to the BaseForm for simpler usage.

Reply:
I have three different forms with three different setups.  I'm not exactly sure what one would extract.  I could create routines to retrieve an organization or a group or a project and call them as needed but I don't think that really helps significantly.

Reply:
BaseForm isn't really the correct place because the checks are only common for the project forms.  Created a new BaseProjectForm and changed both the project forms to derive from that

Reply:
I just meant that maybe it would make sense to have something like:

```csharp
AddError(message)
```

where `AddError` is defined on the base class as:

```
void AddError(string message) {
  var error = new Error(message);

  Errors.add(error);
}
```

or something like that -- just to remove some of the boilerplate

Reply:
Added that as well